### DISCUSSION TITLE
Deploying Image without internet access.

### DISCUSSION THREAD

**Discussion by dpweldo** [open-webui/open-webui#2631] [Q&A] [answered]
I have a cluster running Kubernetes, but it is connected to a private network with no internet access. The image seems to get hung up after deployment when it tries to update from huggingface. Is there an offline version of the image that I can run, or have the dependencies downloaded in the cache so I can re-image the repository for an offline  version?

```sh
Loading WEBUI_SECRET_KEY from file, not provided as an environment variable.
Generating WEBUI_SECRET_KEY
Loading WEBUI_SECRET_KEY from .webui_secret_key
Cannot determine model snapshot path: Cannot find an appropriate cached snapshot folder for the specified revision on the local disk and outgoing traffic has been disabled. To enable repo look-ups and downloads online, pass 'local_files_only=False' as input.
Traceback (most recent call last):
  File "/app/backend/apps/rag/utils.py", line 396, in get_model_path
    model_repo_path = snapshot_download(**snapshot_kwargs)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/huggingface_hub/utils/_validators.py", line 114, in _inner_fn
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/huggingface_hub/_snapshot_download.py", line 220, in snapshot_download
    raise LocalEntryNotFoundError(
huggingface_hub.utils._errors.LocalEntryNotFoundError: Cannot find an appropriate cached snapshot folder for the specified revision on the local disk and outgoing traffic has been disabled. To 
enable repo look-ups and downloads online, pass 'local_files_only=False' as input.
```

```sh
OSError: We couldn't connect to 'https://huggingface.co' to load this file, couldn't find it in the cached files and it looks like sentence-transformers/all-MiniLM-L6-v2 is not the path to a directory containing a file named config.json.
Checkout your internet connection or see how to run the library in offline mode at 'https://huggingface.co/docs/transformers/installation#offline-mode'
```

**Comment by justinh-rahb** [comment]
ensure that `WHISPER_MODEL_AUTO_UPDATE`, `RAG_EMBEDDING_MODEL_AUTO_UPDATE`, and `RAG_RERANKING_MODEL_AUTO_UPDATE` are all set to `False`.

**Reply by dpweldo** [reply]
Thanks for the assistance. I set the parameters you mentioned to `false` like so: 
```py
WHISPER_MODEL_AUTO_UPDATE = False
```
and also like this
```py
WHISPER_MODEL_AUTO_UPDATE = (
    os.environ.get("WHISPER_MODEL_AUTO_UPDATE", "").lower() == "false"
)
```
The images were unable to work and get caught on the `snapshot_dowload` function in the `get_model_path` function. Any ideas on how to bypass this part? 

This is the error log from the container:

```sh
Cannot determine model snapshot path: Cannot find an appropriate cached snapshot folder for the specified revision on the local disk and outgoing traffic has been disabled. To enable repo look-ups and downloads online, pass 'local_files_only=False' as input.
Traceback (most recent call last):
  File "/app/backend/apps/rag/utils.py", line 396, in get_model_path
    model_repo_path = snapshot_download(**snapshot_kwargs)
                      ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/huggingface_hub/utils/_validators.py", line 114, in _inner_fn
    return fn(*args, **kwargs)
           ^^^^^^^^^^^^^^^^^^^
  File "/usr/local/lib/python3.11/site-packages/huggingface_hub/_snapshot_download.py", line 220, in snapshot_download
    raise LocalEntryNotFoundError(
huggingface_hub.utils._errors.LocalEntryNotFoundError: Cannot find an appropriate cached snapshot folder for the specified revision on the local disk and outgoing traffic has been disabled. To 
enable repo look-ups and downloads online, pass 'local_files_only=False' as input.
No sentence-transformers model found with name sentence-transformers/all-MiniLM-L6-v2. Creating a new one with MEAN pooling.
```

**Reply by justinh-rahb** [reply]
Exactly how did you set those? They're to be used as environment variables, so for example with Docker:
```bash
docker run -d -e WHISPER_MODEL_AUTO_UPDATE=False --add-host=host.docker.internal:host-gateway -v open-webui:/app/backend/data --restart unless-stopped --name open-webui ghcr.io/open-webui/open-webui:main
```
Or if you're using direct install, make sure those are set appropriately in the `.env` file.

**Reply by justinh-rahb** [reply]
Another possible issue: are you located in a country that blocks Hugging Face?

**Reply by dpweldo** [reply]
Will the image not start without access to huggingface? Is it possible to include that portion when building the image?

I am trying to deploy on a cluster for our research group, which will have no internet access when it is deployed. With the goal to put it on our registry so it can be deployed and for all of the users. I hoping to modify the image or pass variables so it starts with no internet access.

**Reply by justinh-rahb** [reply, marked as answer]
We've identified the cause. Because of the way volumes are handled upon creation in Kubernetes, the contents of our container image that normally would be pre-loaded in `/app/backend/data` do not get copied to the persistent volume, necessitating the download of the STT and RAG models. On Docker using Docker volumes, this content would be copied upon creation, whereas if you used bind-mount it would not, same as how the PVC does not get populated.

**Reply by dpweldo** [reply]
Oh thanks! I'll try to make a fix to that and let you know how it works out.

**Reply by dpweldo** [reply]
Update: I created a side car container to act as the volume, but it seem to get overwritten on container startup. I am using the helm deployment from this repo https://github.com/open-webui/helm-charts. Any suggestions on how to get the volume to copy over to the persistent volume correctly?

**Reply by justinh-rahb** [reply]
Unfortunately no, I know very little about k8s, our community maintains the helm charts. @0xThresh would be good to ask, he contributed a large portion of that helm repo.

**Reply by james-westbrook** [reply]
@dpweldo have you tried using the `existingClaim` parameter? That would cause your deployment to use an existing PVC instead of building a new one on startup. You can see that in the values file here: https://github.com/open-webui/helm-charts/blob/main/charts/open-webui/values.yaml#L44

**Reply by dpweldo** [reply]
I was able to get it to work by deleting the volume in the deployment after installing it with Helm. It seems that the PV overwrites the data that should be in the `/app/backend/data`, causing it to not be found by the image on start-up. I'll see if I can modify the yamls so I can still have some form of PV and deploy it without having to remove the volume when it is create.

**Reply by DejiCodes** [reply]
Hi, were you able to find a solution to this?
